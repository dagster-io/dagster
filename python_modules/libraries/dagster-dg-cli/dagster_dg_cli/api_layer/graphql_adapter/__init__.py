"""GraphQL adapter layer for Dagster Plus CLI."""
