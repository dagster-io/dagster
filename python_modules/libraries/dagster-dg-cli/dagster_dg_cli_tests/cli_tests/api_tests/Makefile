.PHONY: rerecord test clean

rerecord:
	dagster-dev dg-api-record asset
	dagster-dev dg-api-record deployment
	pytest . --snapshot-update -v

test:
	pytest . -v

clean:
	rm -rf asset_tests/recordings/
	rm -rf deployment_tests/recordings/
