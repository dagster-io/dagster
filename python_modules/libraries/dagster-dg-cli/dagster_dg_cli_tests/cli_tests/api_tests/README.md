# API Test Infrastructure

## QUICKSTART

1. **Add your command to `scenarios.yaml`:**

   ```yaml
   my_new_test:
     command: "dg api asset list --json"
   ```

2. **Record the GraphQL response:**

   ```bash
   dagster-dev dg-api-record asset --recording my_new_test
   ```

3. **Generate snapshots:**
   ```bash
   pytest api_tests/asset_tests/ --snapshot-update
   ```

That's it! Your test is now ready. ðŸŽ‰

### Running tests:

```bash
# Run all tests for a domain
pytest api_tests/asset_tests/

# Update snapshots when output changes
pytest api_tests/asset_tests/ --snapshot-update
```

## Overview

This directory contains the test infrastructure for Dagster Plus CLI API commands. The system uses:

- **GraphQL response recording** from real APIs
- **Syrupy snapshot testing** for output validation
- **Dynamic test discovery** from YAML configuration files

## Directory Structure

```
api_tests/
â”œâ”€â”€ asset_tests/              # Asset API domain tests
â”‚   â”œâ”€â”€ scenarios.yaml        # Test scenario definitions
â”‚   â”œâ”€â”€ recordings/             # Recorded GraphQL responses
â”‚   â”‚   â”œâ”€â”€ success_single_asset/
â”‚   â”‚   â”‚   â””â”€â”€ 01_response.json
â”‚   â”‚   â””â”€â”€ error_asset_not_found/
â”‚   â”‚       â””â”€â”€ 01_response.json
â”‚   â”œâ”€â”€ __snapshots__/        # Syrupy test snapshots
â”‚   â””â”€â”€ test_business_logic.py
â”œâ”€â”€ deployment_tests/         # Deployment API domain tests
â”‚   â””â”€â”€ ...                   # Same structure as asset_tests
â””â”€â”€ test_dynamic_command_execution.py  # Shared test runner
```

## Key Concepts

### Scenarios

Each test scenario is defined in `scenarios.yaml` with:

- **Scenario name**: Descriptive identifier (e.g., `success_single_asset`)
- **Command**: The CLI command to test (e.g., `dg api asset list --json`)

### Recordings

GraphQL responses are recorded from live APIs and stored as JSON files:

- Located in `recordings/{scenario_name}/01_response.json`
- Multiple responses supported with numbered files (01, 02, etc.)
- Automatically created by `dg-api-record` command

### Snapshots

Output validation uses syrupy snapshots:

- Automatically generated on first test run
- Stored in `__snapshots__/` directories
- Both JSON and text output formats supported

## Common Tasks

### Add a new API domain

1. **Create domain directory:**

   ```bash
   mkdir api_tests/my_domain_tests
   mkdir api_tests/my_domain_tests/recordings
   ```

2. **Create scenarios.yaml:**

   ```yaml
   # api_tests/my_domain_tests/scenarios.yaml
   success_list:
     command: "dg api my_domain list --json"
   ```

3. **Record responses:**

   ```bash
   dagster-dev dg-api-record my_domain
   ```

4. **Run tests to generate snapshots:**
   ```bash
   pytest api_tests/my_domain_tests/ --snapshot-update
   ```

### Update existing tests

When CLI output changes:

```bash
# Re-run tests and update snapshots
pytest api_tests/asset_tests/ --snapshot-update

# Review changes
git diff api_tests/asset_tests/__snapshots__/
```

When API responses change:

```bash
# Re-record responses
dagster-dev dg-api-record asset --recording success_single_asset

# Update snapshots
pytest api_tests/asset_tests/ --snapshot-update
```

### Debug test failures

```bash
# Run tests with verbose output
pytest api_tests/asset_tests/ -v

# Run specific test scenario
pytest api_tests/asset_tests/ -k "success_single_asset"

# Show detailed diff on snapshot mismatch
pytest api_tests/asset_tests/ --snapshot-diff
```

## Command Reference

### dg-api-record

Record GraphQL responses for test scenarios:

```bash
# Record all scenarios in a domain
dagster-dev dg-api-record asset

# Record specific scenario
dagster-dev dg-api-record asset --recording success_single_asset

# Available domains
dagster-dev dg-api-record [asset|deployment|run|...]
```

### pytest options

```bash
--snapshot-update    # Accept current output as new baseline
--snapshot-diff      # Show detailed diff on mismatch
-k PATTERN          # Run only tests matching pattern
-v                  # Verbose output
```

## Test Types

### Business Logic Tests (`test_business_logic.py`)

- Test data processing functions in isolation
- Use recorded responses as input
- Validate transformed output with snapshots

### Dynamic Command Execution Tests (`test_dynamic_command_execution.py`)

- Automatically discover and run all scenarios
- Mock GraphQL client with recorded responses
- Validate CLI output against snapshots

### REST Compliance Tests (`test_rest_compliance.py`)

- Validate API method naming conventions
- Check type signatures and response patterns
- Ensure consistency across API domains

## Best Practices

1. **Descriptive scenario names**: Use clear names like `success_multiple_assets` or `error_not_found`
2. **Test both success and error cases**: Include `error_*` scenarios for failure paths
3. **Keep recordings minimal**: Record only the data needed for tests
4. **Review snapshot changes**: Always inspect diffs before committing
5. **Update recordings when APIs change**: Re-record to match current behavior

## Troubleshooting

### "Exhausted responses" error

The test tried to make more GraphQL calls than recorded. Re-record the scenario.

### Snapshot mismatches

Either the output changed (update snapshots) or there's a regression (fix the code).

### Missing recordings

Run `dg-api-record` to create the recording directory and record responses.

### Tests not discovering scenarios

Ensure `scenarios.yaml` exists and has valid YAML syntax.

## Architecture Notes

The test infrastructure separates concerns:

- **Recording**: Captures real API responses once
- **Mocking**: Replays recorded responses during tests
- **Validation**: Snapshots verify output consistency
- **Discovery**: Dynamic test generation from YAML

This design enables:

- Fast test execution (no network calls)
- Reliable tests (deterministic responses)
- Easy updates (re-record or update snapshots)
- Broad coverage (all scenarios tested automatically)
