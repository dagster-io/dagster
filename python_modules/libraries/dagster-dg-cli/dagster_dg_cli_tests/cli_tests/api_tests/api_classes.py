"""Registry of all API classes for REST compliance testing."""

from dagster_dg_cli.api_layer.api.asset import DgApiAssetApi
from dagster_dg_cli.api_layer.api.deployments import DgApiDeploymentApi


def get_all_api_classes():
    """Get all API classes to test."""
    return [
        DgApiDeploymentApi,
        DgApiAssetApi,
    ]
