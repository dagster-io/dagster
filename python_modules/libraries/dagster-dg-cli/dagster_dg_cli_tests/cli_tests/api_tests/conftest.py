"""Shared test utilities for API testing."""

from typing import TYPE_CHECKING
from unittest.mock import Mock

import pytest
from dagster_shared.plus.config import DagsterPlusCliConfig
from syrupy.extensions.amber import AmberSnapshotExtension

if TYPE_CHECKING:
    from syrupy.location import PyTestLocation
    from syrupy.types import SnapshotIndex


class ApiSnapshotExtension(AmberSnapshotExtension):
    """Custom snapshot extension for API tests.

    This extension ensures consistent snapshot naming by:
    1. Removing any existing parameterized test suffixes to avoid duplicate snapshots
    2. Applying a standardized index format that ensures all snapshots follow the pattern test_name[index]

    Without this extension, Syrupy's default naming might create inconsistent snapshot names
    or fail to handle indexing properly for these API tests.
    """

    @classmethod
    def get_snapshot_name(  # pyright: ignore[reportIncompatibleMethodOverride]
        cls,
        *,
        index: "SnapshotIndex",
        test_location: "PyTestLocation",
    ) -> str:
        snapshot_name = test_location.snapshot_name

        # Remove parameterized test suffixes to avoid duplicate snapshots
        variant_start_index = snapshot_name.find("[")
        if variant_start_index != -1:
            snapshot_name = snapshot_name[:variant_start_index]

        return f"{snapshot_name}[{index}]"


@pytest.fixture
def mock_dagster_plus_config():
    """Mock DagsterPlusCliConfig for testing."""
    config = Mock(spec=DagsterPlusCliConfig)
    config.org_name = "test-org"
    config.api_token = "test-token"
    config.endpoint = "https://test.dagster.cloud/graphql"
    return config


@pytest.fixture(autouse=True)
def disable_telemetry(monkeypatch, tmp_path):
    """Disable Dagster telemetry for API tests to prevent output interference."""
    # Set up a temporary DAGSTER_HOME with telemetry disabled
    dagster_home = tmp_path / "dagster_home"
    dagster_home.mkdir()

    # Create dagster.yaml with telemetry disabled
    dagster_yaml = dagster_home / "dagster.yaml"
    dagster_yaml.write_text("telemetry:\n  enabled: false\n")

    # Set the DAGSTER_HOME environment variable
    monkeypatch.setenv("DAGSTER_HOME", str(dagster_home))


@pytest.fixture
def snapshot(snapshot):
    """Configure snapshot fixture for API tests."""
    return snapshot.use_extension(ApiSnapshotExtension)
