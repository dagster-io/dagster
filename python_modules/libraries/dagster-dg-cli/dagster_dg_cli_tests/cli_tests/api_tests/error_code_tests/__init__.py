"""Error code testing for Dagster API CLI.

This package contains comprehensive tests for structured error handling across all API domains.
Tests validate proper error mapping from GraphQL to REST conventions, error response formats,
and exit code behavior.
"""
