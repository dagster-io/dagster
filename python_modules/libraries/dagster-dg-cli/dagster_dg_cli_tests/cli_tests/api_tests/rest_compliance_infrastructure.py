"""Infrastructure for REST compliance testing of Dagster Plus API interfaces."""

import importlib
import inspect
import pkgutil
from functools import cache
from typing import Union, get_args, get_origin

from pydantic import BaseModel

# Primitive types allowed in REST APIs
ALLOWED_PRIMITIVES = (str, int, float, bool, type(None))

# REST method prefixes and their meanings
REST_PREFIXES = {
    "list_": "GET collection",
    "get_": "GET single resource",
    "create_": "POST resource",
    "update_": "PUT/PATCH resource",
    "delete_": "DELETE resource",
    "action_": "POST custom action",
}


@cache
def get_pydantic_models_from_schemas():
    """Discover all Pydantic models from the schemas package."""
    models = set()
    schemas_module = importlib.import_module("dagster_dg_cli.api_layer.schemas")

    # Iterate through all modules in the schemas package
    for _, module_name, _ in pkgutil.iter_modules(schemas_module.__path__):
        full_module_name = f"dagster_dg_cli.api_layer.schemas.{module_name}"
        module = importlib.import_module(full_module_name)
        for name, obj in inspect.getmembers(module):
            if inspect.isclass(obj) and issubclass(obj, BaseModel) and obj != BaseModel:
                models.add(name)

    return frozenset(models)  # Return frozenset for hashability with @cache


def is_pydantic_model(type_hint) -> bool:
    """Check if a type hint is a Pydantic BaseModel subclass."""
    # Handle string annotations (forward references)
    if isinstance(type_hint, str):
        # Check against discovered Pydantic models
        return type_hint in get_pydantic_models_from_schemas()

    # Handle Optional types
    if get_origin(type_hint) in (Union, type(Union)):
        # Check all args in the Union
        args = get_args(type_hint)
        # Filter out NoneType
        non_none_args = [arg for arg in args if arg is not type(None)]
        if len(non_none_args) == 1:
            return is_pydantic_model(non_none_args[0])
        return False

    # Check if it's a class and subclass of BaseModel
    return inspect.isclass(type_hint) and issubclass(type_hint, BaseModel)


def is_allowed_type(type_hint) -> bool:
    """Check if a type hint is allowed (primitive or Pydantic model)."""
    # Handle Optional types
    if get_origin(type_hint) in (Union, type(Union)):
        args = get_args(type_hint)
        return all(is_allowed_type(arg) for arg in args)

    # Check if it's a primitive or Pydantic model
    return type_hint in ALLOWED_PRIMITIVES or is_pydantic_model(type_hint)
