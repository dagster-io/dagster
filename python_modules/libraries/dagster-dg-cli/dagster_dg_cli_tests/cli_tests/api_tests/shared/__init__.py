"""Shared utilities for API test recordings."""
