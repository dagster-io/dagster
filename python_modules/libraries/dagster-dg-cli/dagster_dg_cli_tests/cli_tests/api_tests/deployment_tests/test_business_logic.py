"""Test deployment business logic functions without mocks.

These tests focus on testing pure functions that process data without requiring
GraphQL client mocking or external dependencies.
"""

import json

from dagster_dg_cli.api_layer.graphql_adapter.deployment import process_deployments_response
from dagster_dg_cli.api_layer.schemas.deployment import Deployment, DeploymentList, DeploymentType
from dagster_dg_cli.cli.api.formatters import format_deployments

from dagster_dg_cli_tests.cli_tests.api_tests.test_dynamic_command_execution import (
    load_recorded_graphql_responses,
)


class TestProcessDeploymentsResponse:
    """Test the pure function that processes GraphQL responses."""

    def test_successful_response_processing(self, snapshot):
        """Test processing a successful GraphQL response."""
        response = load_recorded_graphql_responses("deployment", "success_multiple_deployments")[0]
        result = process_deployments_response(response)

        # Snapshot the entire result to capture structure and data
        snapshot.assert_match(result)

    def test_response_processing_with_limit(self, snapshot):
        """Test response processing with a limit applied."""
        response = load_recorded_graphql_responses("deployment", "success_multiple_deployments")[0]
        result = process_deployments_response(response)

        # Snapshot the full result (limit functionality moved to higher level)
        snapshot.assert_match(result)

    def test_empty_response_processing(self, snapshot):
        """Test processing an empty GraphQL response."""
        response = load_recorded_graphql_responses("deployment", "empty_deployments")[0]
        result = process_deployments_response(response)

        # Snapshot empty result
        snapshot.assert_match(result)

    def test_missing_full_deployments_key(self, snapshot):
        """Test processing a response missing the fullDeployments key."""
        malformed_response = {}
        result = process_deployments_response(malformed_response)

        # Snapshot result when key is missing
        snapshot.assert_match(result)


class TestFormatDeployments:
    """Test the deployment formatting functions."""

    def _create_sample_deployment_list(self):
        """Create sample DeploymentList from fixture data."""
        # Load from fixture and convert to domain objects
        response = load_recorded_graphql_responses("deployment", "success_multiple_deployments")[0]
        deployments = [
            Deployment(
                id=dep["deploymentId"],
                name=dep["deploymentName"],
                type=DeploymentType(dep["deploymentType"]),
            )
            for dep in response["fullDeployments"]
        ]
        return DeploymentList(items=deployments, total=len(deployments))

    def test_format_deployments_text_output(self, snapshot):
        """Test formatting deployments as text."""
        deployment_list = self._create_sample_deployment_list()
        result = format_deployments(deployment_list, as_json=False)

        # Snapshot the entire text output
        snapshot.assert_match(result)

    def test_format_deployments_json_output(self, snapshot):
        """Test formatting deployments as JSON."""
        deployment_list = self._create_sample_deployment_list()
        result = format_deployments(deployment_list, as_json=True)

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        parsed = json.loads(result)
        snapshot.assert_match(parsed)
