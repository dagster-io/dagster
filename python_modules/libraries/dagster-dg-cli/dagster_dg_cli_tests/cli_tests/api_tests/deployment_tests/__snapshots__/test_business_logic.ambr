# serializer version: 1
# name: TestFormatDeployments.test_format_deployments_json_output[0]
  dict({
    'items': list([
      dict({
        'id': 6935,
        'name': 'uat',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6936,
        'name': 'excessive',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 84915,
        'name': 'ben-test',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 37,
        'name': 'ecs',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 6934,
        'name': 'staging',
        'type': 'PRODUCTION',
      }),
      dict({
        'id': 80,
        'name': 'prod',
        'type': 'PRODUCTION',
      }),
    ]),
    'total': 6,
  })
# ---
# name: TestFormatDeployments.test_format_deployments_text_output[0]
  '''
  Name: uat
  ID: 6935
  Type: PRODUCTION
  
  Name: excessive
  ID: 6936
  Type: PRODUCTION
  
  Name: ben-test
  ID: 84915
  Type: PRODUCTION
  
  Name: ecs
  ID: 37
  Type: PRODUCTION
  
  Name: staging
  ID: 6934
  Type: PRODUCTION
  
  Name: prod
  ID: 80
  Type: PRODUCTION
  '''
# ---
# name: TestProcessDeploymentsResponse.test_empty_response_processing[0]
  DeploymentList(items=[Deployment(id=6935, name='uat', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=6936, name='excessive', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=84915, name='ben-test', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=37, name='ecs', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=6934, name='staging', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=80, name='prod', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>)], total=6)
# ---
# name: TestProcessDeploymentsResponse.test_missing_full_deployments_key[0]
  DeploymentList(items=[], total=0)
# ---
# name: TestProcessDeploymentsResponse.test_response_processing_with_limit[0]
  DeploymentList(items=[Deployment(id=6935, name='uat', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=6936, name='excessive', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=84915, name='ben-test', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=37, name='ecs', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=6934, name='staging', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=80, name='prod', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>)], total=6)
# ---
# name: TestProcessDeploymentsResponse.test_successful_response_processing[0]
  DeploymentList(items=[Deployment(id=6935, name='uat', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=6936, name='excessive', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=84915, name='ben-test', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=37, name='ecs', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=6934, name='staging', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>), Deployment(id=80, name='prod', type=<DeploymentType.PRODUCTION: 'PRODUCTION'>)], total=6)
# ---
